### PR TITLE
Adding C++20 counting semaphore API

### PR DESCRIPTION
**Pull request by hkaiser**
- adding `hpx::binary_semaphore`
- adding `hpx/counting_semaphore.hpp` that imports it into namespace hpx
- flyby: adding `hpx/condition_variable.hpp` and `hpx/stop_token.hpp` headers for
         conformance. This also imports `lcos::local::conditional_variable`
         into namespace hpx

### CODE REVIEW

**Comment by msimberg** [comment]
@hkaiser I summarized our API discussion from yesterday in #4329. Should we already go ahead and collect the "API headers" into the `include` module in this PR (and #4636, #4635)? I can also collect them afterwards while doing all the other headers.

File: libs/synchronization/include/hpx/semaphore.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    /// A counting semaphore is a semaphore object that models a non - negative
+    /// resource count. A binary semaphore is a semaphore object that has only
+    /// two states.
+    ///
+    /// Class template counting_semaphore maintains an internal counter that is
+    /// initialized when the semaphore is created. The counter is decremented
+    /// when a thread acquires the semaphore, and is incremented when a thread
+    /// releases the semaphore. If a thread tries to acquire the semaphore when
+    /// the counter is zero, the thread will block until another thread
+    /// increments the counter by releasing the semaphore.
```
Comment (msimberg) [inline libs/synchronization/include/hpx/semaphore.hpp:33]: No doxygen?
```suggestion
    // increments the counter by releasing the semaphore.
```

File: libs/synchronization/include/hpx/semaphore.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    template <std::ptrdiff_t LeastMaxValue = PTRDIFF_MAX>
+    using counting_semaphore =
+        hpx::lcos::local::cpp20_counting_semaphore<LeastMaxValue>;
+
+    /// A binary semaphore should be more efficient than the default
+    /// implementation of a counting semaphore with a unit resource count.
```
Comment (msimberg) [inline libs/synchronization/include/hpx/semaphore.hpp:39]: No doxygen?
```suggestion
    // implementation of a counting semaphore with a unit resource count.
```

**Comment by hkaiser** [comment]
> Should we already go ahead and collect the "API headers" into the include module in this PR (and #4636, #4635)?

I thought we will create at least 3 `include` modules, one for each of the planned configurations. Should we go ahead and make that split now?

**Comment by msimberg** [comment]
> > Should we already go ahead and collect the "API headers" into the include module in this PR (and #4636, #4635)?
> 
> I thought we will create at least 3 `include` modules, one for each of the planned configurations. Should we go ahead and make that split now?

Yes, I'd say that's the plan in the end. But if we decide to skip splitting up the main library for 1.5.0 then we could still already have them in the include module. Either way. All I would like is that the headers from #4626 that we can put in place we do put in place for 1.5.0. It doesn't matter too much which module they are in before the main library is split into three.

**Comment by hkaiser** [comment]
This should be good to go now as well.

**Comment by msimberg** [comment]
Thanks @hkaiser. I just rebased this because of a conflict with #4634 in the CMake files. I'll merge this as soon as some of the GH actions builders come back clean.

**Pull request merged**